cmake/developer_package/compile_flags/os_flags.cmake

```diff
@@ -62,6 +62,34 @@ macro(ov_deprecated_no_errors)
     set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${ov_c_cxx_deprecated_no_errors}")
 endmacro()
 
+#
+# ov_deprecated_featurs_no_errors()
+#
+# Don't threat deprecated features of C++ as errors in current scope (directory, function)
+# Defines ov_c_cxx_deprecated_features_no_errors varaible which contains C / C++ compiler flags
+#
+macro(ov_deprecated_featurs_no_errors)
+    if(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+        # show 4996 only for /w4
+        set(ov_c_cxx_deprecated_features_no_errors "")
+    elseif(CMAKE_CXX_COMPILER_ID STREQUAL "Intel")
+        if(WIN32)
+            set(ov_c_cxx_deprecated_features_no_errors "")
+        else()
+            set(ov_c_cxx_deprecated_features_no_errors "")
+        endif()
+    elseif(OV_COMPILER_IS_CLANG OR CMAKE_COMPILER_IS_GNUCXX)
+        set(ov_c_cxx_deprecated_features_no_errors "-Wno-error=deprecated")
```

Comment:
why do we need it? Can we have conditional code depending on C++ standard in C++ code itself?

Reply:
Is mostly related to warning as error for lambda captures `[=]`:
` error: implicit capture of ‘this’ via ‘[=]’ is deprecated in C++20` and suggestion to add explicit `this`.

The `[=]` capture is very often used in transformation, also in some places in CPU plugin and frontends.

To be compatible with cpp20 it should be replaced by `[=, this]`, but it cannot be used for C++11.
Another solution could be to use some macro to replace `=` depends on standard for `=` or `=, this`. 
The `=` could be also replaced by explicit list of captures which should be compatible in all standards.
Both solutions require a lot of files to be update and to avoid it I've just disabled these warnings.

Reply:
maybe we can globally use:
```
if(CMAKE_CXX_STANDARD GREATER_EQUAL XX)
    set(CMAKE_CXX_FLAGS "-Wno-error=deprecated ${CMAKE_CXX_FLAGS}")
endif()
```
?

Reply:
I can use and remove helper cmake macro. I can apply it globally or locally (just let me know what would be better).

Benefit of local use of this warning as error suppression could be that where is not used there still can be an error if non-compliant code used.

This warning is optional for compiler and it looks like `gcc` is emit it only for now but in future in newer standard can break compilation. If we want support wide range of standards then our code should be updated to fix this warning. I think component owners should decide how to fix e.g in plugins, transformations (biggest job) etc.

I can propose to use explicit list where `this` can be captured implicit as this syntax is well supported from cpp11 and give more error-proof code. 

There are some recommendations from SonarQube tool, also:
- [Default capture should not be used](https://rules.sonarsource.com/cpp/RSPEC-3608/?search=lambda)
- [Lambdas that capture "this" should capture everything explicitly](https://rules.sonarsource.com/cpp/RSPEC-5019/?search=lambda)

Reply:
> I can propose to use explicit list where this can be captured implicit as this syntax is well supported from cpp11 and give more error-proof code.

I agree.
Let's create tasks for components to make the code universal across all C++ standards